zjit/src/hir.rs

```diff
@@ -1607,40 +1611,60 @@ impl Function {
         }
     }
 
+    /// Set self.param_types. They are copied to the param types of entry blocks.
+    fn set_param_types(&mut self) {
+        let iseq = self.iseq;
+        let param_size = unsafe { get_iseq_body_param_size(iseq) }.as_usize();
+        let rest_param_idx = if !iseq.is_null() && unsafe { get_iseq_flags_has_rest(iseq) } {
+            let opt_num = unsafe { get_iseq_body_param_opt_num(iseq) };
+            let lead_num = unsafe { get_iseq_body_param_lead_num(iseq) };
+            Some(opt_num + lead_num)
+        } else {
+            None
+        };
+
+        self.param_types.push(types::BasicObject); // self
+        for local_idx in 0..param_size {
+            let param_type = if Some(local_idx as i32) == rest_param_idx {
+                types::ArrayExact // Rest parameters are always ArrayExact
+            } else {
+                types::BasicObject
+            };
+            self.param_types.push(param_type);
+        }
+    }
+
+    /// Copy self.param_types to the param types of entry blocks.
+    fn copy_param_types(&mut self) {
+        for entry_block in self.entry_blocks() {
+            let entry_params = self.blocks[entry_block.0].params.iter();
+            let param_types = self.param_types.iter();
+            assert_eq!(
+                entry_params.len(),
+                param_types.len(),
+                "param types should be initialized before type inference",
+            );
+            for (param, param_type) in std::iter::zip(entry_params, param_types.clone()) {
```

Comment:
```suggestion
            for (param, param_type) in std::iter::zip(entry_params, param_types) {
```
Shouldn't need to clone the iterator

Reply:
thanks, this was a leftover from the version where it was in a loop. https://github.com/ruby/ruby/pull/14653/commits/4a7d0e4321bbb0fc194c819b3295a9c4d12498e9